src/liboslcomp/typecheck.cpp

```diff
@@ -1180,58 +1156,422 @@ ASTfunction_call::typecheck_struct_constructor ()
+                     bool (ASTNode::*checkfunc) (const char *funcname,
+                                                 ASTNode::ref arg,
+                                                 const char *formals,
+                                                 bool coerce))
+        : m_compiler(comp), m_func(func), m_root(root), m_check_arglist(checkfunc) {
     }
 
-    // All that failed, try for a coercible match on everything
-    if (! match && expected != TypeSpec()) {
-        m_typespec = typecheck_all_poly (TypeSpec(), true, false);
-        if (m_typespec != TypeSpec())
+    FunctionSymbol*
+    operator () (TypeSpec expected)
+    {
+        bool match = false;
+        TypeSpec typespec;
+        FunctionSymbol* sym;
+
+        // Look for an exact match, including expected return type
+        std::tie(sym, typespec) = typecheck_polys (expected, false, false);
+        if (typespec != TypeSpec())
+            match = true;
+
+        // Now look for an exact match for arguments, but equivalent return type
+        std::tie(sym, typespec) = typecheck_polys (expected, false, true);
+        if (typespec != TypeSpec())
             match = true;
+
+        // Now look for an exact match on args, but any return type
+        if (! match && expected != TypeSpec()) {
+            std::tie(sym, typespec) = typecheck_polys (TypeSpec(), false, false);
+            if (typespec != TypeSpec())
+                match = true;
+        }
+
+        // Now look for a coercible match of args, exact march on return type
+        if (! match) {
+            std::tie(sym, typespec) = typecheck_polys (expected, true, false);
+            if (typespec != TypeSpec())
+                match = true;
+        }
+
+        // Now look for a coercible match of args, equivalent march on return type
+        if (! match) {
+            std::tie(sym, typespec) = typecheck_polys (expected, true, true);
+            if (typespec != TypeSpec())
+                match = true;
+        }
+
+        // All that failed, try for a coercible match on everything
+        if (! match && expected != TypeSpec()) {
+            std::tie(sym, typespec) = typecheck_polys (TypeSpec(), true, false);
+            if (typespec != TypeSpec())
+                match = true;
+        }
+
+        return match ? sym : nullptr;
+    }
+};
+
+
+
+TypeSpec
+ASTfunction_call::typecheck (TypeSpec expected)
+{
+    typecheck_children ();
+
+    if (is_struct_ctr()) {
+        // Looks like function call, but is actually struct constructor
+        return typecheck_struct_constructor ();
+    }
+
+    // Save the currently choosen symbol for error reporting later
+    FunctionSymbol* poly = func();
+
+    CandidateFunctions candidates(m_compiler, expected, args(), poly);
+    std::tie(m_sym, m_typespec) = candidates.best(this);
+
+    // Check resolution against prior versions of OSL
+    const char* OSL_LEGACY = ::getenv("OSL_LEGACY_FUNCTION_RESOLUTION");
```

Comment:
Is it weird or expensive to do this getenv and evaluation for every function call we typecheck? Maybe it's something we should do once in the OSLCompilerImpl at startup time?

Or maybe it doesn't matter because I think we will probably remove it soon. It's primarily so that we don't have to revert again if a minor problem is found.

Reply:
Whatever the cost, you're right paying it multiple times is stupid.
Can easily move it to a static and incur the hit of a lock only?